### ISSUE TITLE
How far from proper installation with `pip`?

### ISSUE THREAD

**Issue by DiTo97** [DreamingRaven/python-reseal#13]
Hi @DreamingRaven,

First of all, thanks for your preliminary work enabling FHE in machine learning. I had two questions in mind specifically about the longevity and long-term goals of this project, since the last commit was more than nine months ago (not a hit on you, I know myself how difficult it can be to maintain and document a medium-to-large library on your own).

1. I saw the installation instructions, and the preferred installation method seems to be still by running the code inside the published Docker containers due to the difficulties building the bindings with the SEAL library. I know that the team working on SEAL has no intention of actively maintaining Python bindings, as they prefer to focus their efforts on the C++ compatibility. Is this [GitLab repository](https://gitlab.com/deepcypher/python-fhez-seal) the one containing your custom SEAL bindings?
2. As a side note, how far do you think you are from having a proper self-contained installation backend with `pip`? Having working containers is nice because we can play with the package, but being able to install it would be crucial.

I am quite interested in your package, since we are trying to solve a real-world problem in the healthcare domain, where the preservation of privacy is the major concern on the data provider's side. Do you think the package is mature enough to build a complex network? Being very RAM and storage hungry, I am trying to figure out what the requirements of such a network would be.

I know that Intel has also developed its own SEAL [integration for HE in machine learning](https://github.com/IntelAI/he-transformer) with the nGraph compiler (now moved to the OpenVINO toolkit). Are there any major differences between your approach to FHE neural networks and theirs?

Finally, are you still actively working on this project? What are your short- and long-term goals with it?

**Comment by DreamingRaven** [comment]
Hey @DiTo97,
Yes the library currently is handicapped by the bindings! I need to update to the latest pybind11 binding that Huelse https://github.com/Huelse/SEAL-Python has been working on. This is the origin of the bindings that I use, but my bindings are out of data and need a rework to be compatible with the latest MS-SEAL versions as there was an update to one of the required libraries which makes it not compile at the moment (see https://gitlab.com/deepcypher/python-fhez/-/issues/1 which is something I super regret not having tackled yet).

1. https://gitlab.com/deepcypher/python-fhez/-/blob/master/Dockerfile_archlinux#L38 currently actually resolves to https://github.com/DreamingRaven/seal-python which is where the bindings exist. There are what need fixing to make the library build again.
2. If I am honest I think I lack some knowledge with the PIP system to enable me to build this quite complex pybind11 setup straight with PIP. The difficulties again come from the bindings.

Now as for the status of the project, I would say with my PhD finishing in 19 days there is hope for me to have time to come back to it and fix it. However after that point it will probably go into maintenance mode as I am no longer a fan of MS-SEAL, but I want to fix it atleast for nice people like yourself so they have something atleast in python to work with. Instead I am likely to commit more to Darklantern (in Go and based on lattigo). https://gitlab.com/deepcypher/darklantern which is necessary for bootstrapping as last I checked MS-SEAL CKKS bindings did not have any bootstrapping which is critical for much deeper neural networks.

I have to shoot off to a meeting for a few hours I will be back shortly to elaborate more, on the last few questions.

**Comment by DiTo97** [comment]
> Hey @DiTo97, Yes the library currently is handicapped by the bindings! I need to update to the latest pybind11 binding that Huelse https://github.com/Huelse/SEAL-Python has been working on. This is the origin of the bindings that I use, but my bindings are out of data and need a rework to be compatible with the latest MS-SEAL versions as there was an update to one of the required libraries which makes it not compile at the moment (see https://gitlab.com/deepcypher/python-fhez/-/issues/1 which is something I super regret not having tackled yet).
> 
> 1. https://gitlab.com/deepcypher/python-fhez/-/blob/master/Dockerfile_archlinux#L38 currently actually resolves to https://github.com/DreamingRaven/seal-python which is where the bindings exist. There are what need fixing to make the library build again.
> 2. If I am honest I think I lack some knowledge with the PIP system to enable me to build this quite complex pybind11 setup straight with PIP. The difficulties again come from the bindings.
> 
> Now as for the status of the project, I would say with my PhD finishing in 19 days there is hope for me to have time to come back to it and fix it. However after that point it will probably go into maintenance mode as I am no longer a fan of MS-SEAL, but I want to fix it atleast for nice people like yourself so they have something atleast in python to work with. Instead I am likely to commit more to Darklantern (in Go and based on lattigo). https://gitlab.com/deepcypher/darklantern which is necessary for bootstrapping as last I checked MS-SEAL CKKS bindings did not have any bootstrapping which is critical for much deeper neural networks.
> 
> I have to shoot off to a meeting for a few hours I will be back shortly to elaborate more, on the last few questions.

Thank you for the quick reply and being so open about your plans.

I was about to report a very similar error to the one in the GitLab issue, as I just tried to build the Docker image in WSL in order to try out the MNIST example, but the build stopped with that same error you mentioned.

I am no PIP or package management expert myself, but I think that trying to replicate in [setup.py](https://github.com/DreamingRaven/python-fhez/blob/master/setup.py) most of the steps you perform in the Dockerfile should be sufficient for an all-in-one installation, e.g., by [adding custom build steps](https://jichu4n.com/posts/how-to-add-custom-build-steps-and-commands-to-setuppy/) to setup.py.

If I may ask, is there any other reason why you no longer like MS-SEAL, other than the lack of CKKS bindings for bootstrapping?

Take your time with the meeting. I will be waiting for your addition later or tomorrow, in no hurry.

**Comment by DreamingRaven** [comment]
> I was about to report a very similar error to the one in the GitLab issue, as I just tried to build the Docker image in WSL in order to try out the MNIST example, but the build stopped with that same error you mentioned.

Yeah its a real shame I spent so long building it all only for the binding build to break thanks to C++ changes. If I had built the container with an old ubuntu image it would still work / function. That might actually be the most expedient way to patch it for now.

> I am no PIP or package management expert myself, but I think that trying to replicate in [setup.py](https://github.com/DreamingRaven/python-fhez/blob/master/setup.py) most of the steps you perform in the Dockerfile should be sufficient for an all-in-one installation, e.g., by [adding custom build steps](https://jichu4n.com/posts/how-to-add-custom-build-steps-and-commands-to-setuppy/) to setup.py.

I will definitely give this an attempt once my thesis is submitted. If it is that straightforward it would make everyones' life easier for installation.

> If I may ask, is there any other reason why you no longer like MS-SEAL, other than the lack of CKKS bindings for bootstrapping?

So for me bootstrapping is really the linchpin for FHE. The lack of it handicaps any attempt at deep learning that is beyond a few neurons deep. As the data would need to be constantly sent back and forth for re-encryption, whereas with bootstrapping the data can be submitted, processed in its entirety, then sent back processed. This will cut down actual use case time performance by orders of magnitude! I personally also feel more capable of writing the deep learning stack than the FHE stack, so why not use the best possible implementation (in my opinion), that is the most portable and usable in the form of lattigo and write the DL myself!

Subsequently while I actually have a wealth of experience in C++ and python I am not keen on the combo via pybind11. I am also not keen on writing DL directly in C++ as it does not offer the convenience and portability of Go. For instance I cannot use the pybind11 bindings of MS-SEAL with web assembly, it takes a lot of time to compile, and is much more arduous to extend. In Go however I can package, version, distribute, and even serve it to peoples web browsers for local ad-hoc learning which will really open up encrypted deep learning to a lot larger of an audience. While simultaneously making it that much easier for me to extend the functionality! Of course this is my subjective opinion but the niche FHE fills overlaps more with Go than it does with C++. Python would have been ideal for the DL side however. A matter of choosing the poison I guess. Go with lattigo, C++ with SEAL, or python with deep learning, then finding some way to attain the missing functionality. 

Also I would note there may be another project that you might be interested in that does what I have done in a very very similar way: https://github.com/microsoft/EVA and is microsofts own python implementation. Just in case you did not stumble upon it as opposed to intel nGraph which as far as I am aware is not supported anymore, and I have not looked into openVINO from a passing glance I did not see FHE on their advertisement page at-least.

But long story short, its tricky. There is no best choice, its just what extents we are willing to go through to get encrypted deep learning to more people.

> I am quite interested in your package, since we are trying to solve a real-world problem in the healthcare domain, where the preservation of privacy is the major concern on the data provider's side. Do you think the package is mature enough to build a complex network? Being very RAM and storage hungry, I am trying to figure out what the requirements of such a network would be.

I would be quite interested in what sort of issue this is, although I realize you probably cant elaborate. Just be aware a gotcha with FHE is training, as you cant mix keys, and if the outputs are encrypted with the same keys as the inputs you can never discern when exactly to stop training (along with inevitably encrypting the weights of the model during backprogation with the very same key). FHE I would say is the most hardcore of the privacy preserving machine learning techniques, and there is still a heck of a lot of research that needs to go into it to make it viable at scale, what can take less than a second in plaintext can take minutes encrypted. If you can I would advise to go with a less extreme PPML technique like federation / distributed computing, or differential privacy. But that is all dependent on your threat model, it may very well be the case that FHE is the only possible solution but it is a very very difficult one.

Just so you know, from my perspective it is the most secure, most provable, of any of the PPML techniques, and is not vulnerable to the same attacks by [evil maids](https://en.wikipedia.org/wiki/Evil_maid_attack) or quantum computers. It may also offer an off ramp for things like GDPR. Which is why I am trying to push encrypted deep learning where I can. However I also try to clarify where it does have flaws like in training and time performance as if I highlight them, they are more likely to be widely known and thus tackled by someone smarter than myself who might then be interested in solving them.

Feel free to ask any questions you might have, I do love this field, I could talk about it all day!

**Comment by DiTo97** [comment]
> Yeah its a real shame I spent so long building it all only for the binding build to break thanks to C++ changes. If I had built the container with an old ubuntu image it would still work / function. That might actually be the most expedient way to patch it for now.

By chance, do you know which old Ubuntu images would work to build the image with?

> I will definitely give this an attempt once my thesis is submitted. If it is that straightforward it would make everyones' life easier for installation.

I mean, it should be that easy, but I cannot promise it ahaha.

I have seen enough fancy stuff in the setup.py of even famous projects to have an intuition that this should be the way to go, although it will probably work only on Linux for now, if you replicate the contents of the Dockerfiles to the letter.

> So for me bootstrapping is really the linchpin for FHE. The lack of it handicaps any attempt at deep learning that is beyond a few neurons deep. As the data would need to be constantly sent back and forth for re-encryption, whereas with bootstrapping the data can be submitted, processed in its entirety, then sent back processed. This will cut down actual use case time performance by orders of magnitude! I personally also feel more capable of writing the deep learning stack than the FHE stack, so why not use the best possible implementation (in my opinion), that is the most portable and usable in the form of lattigo and write the DL myself!

If you put it that way, (lack of) bootstrapping seems to be the true deal breaker for FHE deep learning.

I have only scratched the surface of homomorphic encryption, but the biggest takeaway for now is that the compute and memory requirements of an FHE network are far greater than those of an unencrypted network.

Therefore, anything that can reduce the computational load is absolutely critical. Moreso, if it does it by orders of magnitude.
 
> Subsequently while I actually have a wealth of experience in C++ and python I am not keen on the combo via pybind11. I am also not keen on writing DL directly in C++ as it does not offer the convenience and portability of Go. For instance I cannot use the pybind11 bindings of MS-SEAL with web assembly, it takes a lot of time to compile, and is much more arduous to extend. In Go however I can package, version, distribute, and even serve it to peoples web browsers for local ad-hoc learning which will really open up encrypted deep learning to a lot larger of an audience. While simultaneously making it that much easier for me to extend the functionality! Of course this is my subjective opinion but the niche FHE fills overlaps more with Go than it does with C++. Python would have been ideal for the DL side however. A matter of choosing the poison I guess. Go with lattigo, C++ with SEAL, or python with deep learning, then finding some way to attain the missing functionality.
> 
> Also I would note there may be another project that you might be interested in that does what I have done in a very very similar way: https://github.com/microsoft/EVA and is microsofts own python implementation. Just in case you did not stumble upon it as opposed to intel nGraph which as far as I am aware is not supported anymore, and I have not looked into openVINO from a passing glance I did not see FHE on their advertisement page at-least.
> 
> But long story short, its tricky. There is no best choice, its just what extents we are willing to go through to get encrypted deep learning to more people.

I am more of a deep learning guy myself, with tons of experience in C++ and Python, but Pybind11 is difficult. There is no other way to put it, because it certainly helps bridge the gap between the two worlds, but it is a framework by itself, and like any framework, it has its own rules, which often are not related to the underlying programming language.

Pybind11 is probably one of the few instances where I have made extensive use of AI coding assistants (and for the better).

I have very limited knowledge of Go, though, so I can't really comment on that. One thing that I can say is that each programming language has strengths and weaknesses, and it fills its own niche. If you find yourself more comfortable with Go and Lattigo, and its role in the FHE community, than we already have our answer I guess, so I will keep on eye on that.

I would say it is even easier to [bridge Python and Go](https://www.ardanlabs.com/blog/2020/06/python-go-grpc.html), than it is Python and C++, so as long as a great library is being developed, the chosen programming language is not as discriminating as one might think.

I still can see having Golang handle the FHE side of thins and Python the deep learning's.

About the EVA project, I actually did notice it before yours, but I guess it lacks the whole deep learning stack on top. Moreover, judging from the installation instructions it still quite tricky to setup, so not really suitable for enterprise.

Also, I do not think EVA solves your concerns with MS-SEAL, does it?

About nGraph, and its transition to OpenVINO, I did find it in [OpenVINO's docs](https://docs.openvino.ai/latest/api/ie_python_api/_autosummary/ngraph.html), so I am not sure if it has been dismantled or not. Nonetheless, I know nothing about, so I would have to try it out before being able to make any comments.

> I would be quite interested in what sort of issue this is, although I realize you probably cant elaborate. Just be aware a gotcha with FHE is training, as you cant mix keys, and if the outputs are encrypted with the same keys as the inputs you can never discern when exactly to stop training (along with inevitably encrypting the weights of the model during backprogation with the very same key). FHE I would say is the most hardcore of the privacy preserving machine learning techniques, and there is still a heck of a lot of research that needs to go into it to make it viable at scale, what can take less than a second in plaintext can take minutes encrypted. If you can I would advise to go with a less extreme PPML technique like federation / distributed computing, or differential privacy. But that is all dependent on your threat model, it may very well be the case that FHE is the only possible solution but it is a very very difficult one.

I cannot disclose that much, as you imagined. Let's just say that we have a use case where the data cannot absolutely leave the data provider's servers, unless encrypted, and the data provider has little to no compute to do anything on-prem.

I already proposed differential privacy, but I am waiting for confirmation, since I am more familiar with that paradigm. I completely understand your warning about using the same keys in FHE training, but how would you avoid that?

From a technical standpoint, assuming the data comes encrypted with some key in the network, how would you make the outputs come out encrypted with some other key before back propagating the information?

> Just so you know, from my perspective it is the most secure, most provable, of any of the PPML techniques, and is not vulnerable to the same attacks by [evil maids](https://en.wikipedia.org/wiki/Evil_maid_attack) or quantum computers. It may also offer an off ramp for things like GDPR. Which is why I am trying to push encrypted deep learning where I can. However I also try to clarify where it does have flaws like in training and time performance as if I highlight them, they are more likely to be widely known and thus tackled by someone smarter than myself who might then be interested in solving them.

Thank you so much for the wealth of information you have let out. It has been absolutely useful.  
I will close the issue for now, but I will re-open it in the future if I have more questions.

Good luck with your PhD thesis defense!

**Comment by DreamingRaven** [comment]
> By chance, do you know which old Ubuntu images would work to build the image with?

I believe ubuntu 18 as that is what Huelse originally used.

> I already proposed differential privacy, but I am waiting for confirmation, since I am more familiar with that paradigm. I completely understand your warning about using the same keys in FHE training, but how would you avoid that?

So the standard approach currently is different for training and inference. As I mentioned before encrypted deep learning does not work well for training, as you eventually have to decrypt to calculate some form of loss for the neural net to backprop with, and you also need this loss so you can tell when to stop training. You can backprop with an encrypted value but when you update the weights they will become encrypted. So you tend to do backprop and weightupdate in plaintext. However now thinking about it you can probably just do Adam in plaintext and call it a day to avoid the neural network weights becoming cyphertexts.

For inference it is far easier as you don't need to backprop so you just go forward and spit out encrypted answers using the already trained network.

> From a technical standpoint, assuming the data comes encrypted with some key in the network, how would you make the outputs come out encrypted with some other key before back propagating the information?

You don't, the output is a transformation of the input. At the end of the day applying a neural network should be almost like applying a super long polynomial (ignoring activation functions for now). Thus what you put in, is what you get out, just added, multiplied, and subtracted to.

> Also, I do not think EVA solves your concerns with MS-SEAL, does it?

No, it doesn't unfortunately. But if you were set on MS-SEAL thats probably one of the few ways to go for python integration.

> Good luck with your PhD thesis defense!

And thanks for the luck!

**Comment by DiTo97** [comment]
> So the standard approach currently is different for training and inference. As I mentioned before encrypted deep learning does not work well for training, as you eventually have to decrypt to calculate some form of loss for the neural net to backprop with, and you also need this loss so you can tell when to stop training. You can backprop with an encrypted value but when you update the weights they will become encrypted. So you tend to do backprop and weightupdate in plaintext. However now thinking about it you can probably just do Adam in plaintext and call it a day to avoid the neural network weights becoming cyphertexts.

The last bit on Adam in plaintext is what was behind my confusion on the encryption of the forward and backward passes. Rather, why we could not just do it and forget about the whole *weights-turning-into-cyphertexts* mess?

The way I see it, the forward-backward cycle of FHE neural networks should look something like this:

```mermaid
stateDiagram-v2
    direction LR

    A: Batch plaintext
    B: Input plaintext
    C: Target plaintext
    D: Input cyphertext
    E: FHE  neural network
    F: Output cyphertext
    G: Output plaintext
    H: Loss

    A --> B
    A --> C
    B --> D: encrypt
    D --> E: submit
    E --> F: forward
    F --> G: decrypt

    state join_state <<join>>
    C --> join_state
    G --> join_state

    join_state --> H: compute loss
    H --> E: backward
```
Am I missing something?

**Comment by DreamingRaven** [comment]
No that looks like a solid representation of the whole thing. https://python-fhez.readthedocs.io/en/dev/examples/constelation.html this for instance is the directed (arrows are missing due to limitation of plotter) acyclic graph of a FHE compatible neural network computation. You start from x and go to y for forward pass. Then reverse the direction for backward pass. This one also includes an explicit decyption stage so you can see when it is exactly decrypted. Or this one for Fashion-MNIST https://python-fhez.readthedocs.io/en/dev/examples/fashion-mnist.html which shows the same thing with an ever so slightly altered network architecture.

But yeah in my implementation the entire backward pass is done in plaintext. Although often if you can go forward in an FHE compatible manner you can often backprop too https://python-fhez.readthedocs.io/en/dev/activations.html (blue means compatible, red means incompatible) all the way up until adam https://python-fhez.readthedocs.io/en/dev/optimisers.html which is just not going to work with FHE for now atleast.
